Neos.Flow/Classes/ResourceManagement/Streams/PackageStreamWrapper.php

```diff
@@ -0,0 +1,526 @@
+     */
+    public function resourceStat()
+    {
+        return fstat($this->handle);
+    }
+
+    /**
+     * Retrieve information about a file.
+     *
+     * This method is called in response to all stat() related functions.
+     *
+     * $flags can hold one or more of the following values OR'd together:
+     *  STREAM_URL_STAT_LINK
+     *     For resources with the ability to link to other resource (such as an
+     *     HTTP Location: forward, or a filesystem symlink). This flag specified
+     *     that only information about the link itself should be returned, not
+     *     the resource pointed to by the link. This flag is set in response to
+     *     calls to lstat(), is_link(), or filetype().
+     *  STREAM_URL_STAT_QUIET
+     *     If this flag is set, your wrapper should not raise any errors. If
+     *     this flag is not set, you are responsible for reporting errors using
+     *     the trigger_error() function during stating of the path.
+     *
+     * Note: The stat() call is silenced through the shut-up operator because this method would issue a warning if the
+     *       file does not exist - but file_exists() will call pathStat() in order to check exactly that. So without
+     *       the "@" operator it wouldn't be possible to run file_exists() on a resource without issuing a warning and
+     *       the resulting exception.
+     *
+     * @param string $path The file path or URL to stat. Note that in the case of a URL, it must be a :// delimited URL. Other URL forms are not supported.
+     * @param integer $flags Holds additional flags set by the streams API.
+     * @return array Should return as many elements as stat() does. Unknown or unavailable values should be set to a rational value (usually 0).
+     */
+    public function pathStat($path, $flags)
+    {
+        $evaluatedPath = $this->evaluatePackagePath($path);
+        if ($evaluatedPath === false) {
+            return false;
+        }
+        return @stat($evaluatedPath);
+    }
+
+    /**
+     * Evaluates the absolute path and filename of the package file specified
+     * by the given path.
+     *
+     * @param string $requestedPath
+     * @param boolean $checkForExistence Whether a (non-hash) path should be checked for existence before being returned
+     * @return mixed The full path and filename or false if the file doesn't exist
+     * @throws \InvalidArgumentException|ResourceException
+     */
+    protected function evaluatePackagePath($requestedPath, $checkForExistence = true)
+    {
+        $requestPathParts = explode('://', $requestedPath, 2);
+        if ($requestPathParts[0] !== self::SCHEME) {
+            throw new \InvalidArgumentException('The ' . __CLASS__ . ' only supports the \'' . self::SCHEME . '\' scheme.', 1256052544);
+        }
+
+        if (!isset($requestPathParts[1])) {
+            return false;
+        }
+
+        if (preg_match('/(^|\/)\\.\\.\//us', $requestPathParts[1])) {
```

Comment:
What do you think about resolving the $resourceUri path with realpath and check if it is inside of the current package. That could save us this regex (which confuses me a bit ^^) ;)

On the other hand it is all tested so fine either way - thanks ;)

Reply:
but using realpath woud be an unnecessary i/o maybe we can use a fake version of realpath https://www.php.net/manual/de/function.realpath.php
that operates on the package:// path to see if it traverses outward

... or i start understanding the regex ;)

Reply:
At https://stackoverflow.com/questions/611520/preventing-directory-traversal-with-web-facing-application-are-regular-express someone wrote

>  using a Regex based protection for such a scenario is like wearing a T-shirt against a hurricane

While I wouldn't put it that harsh, I agree that using `realpath` is probably a good idea ;)

Reply:
How would you compare that? Realpath will just convert `..` without complaining

Reply:
i would check if the realpath path starts with $package->getPackagePath() if not - someone used dark magic to traverse outwards

-> given that $package->getPackagePath() returns a realpath itself

Reply:
+100 for checking with `realpath` here! Security over performance, definitely. If it makes a difference at all, given that regex handling isn't the fastest thing on earth anyway.

Reply:
we can still have a "fast" path when no `..` are included in the string ^^